src/NHapi.Base/Parser/XMLParser.cs

```diff
@@ -25,796 +25,828 @@
+        /// </summary>
+        private bool EncodePrimitive(IPrimitive datatypeObject, XmlElement datatypeElement)
+        {
+            var hasValue = false;
+            if (datatypeObject.Value != null && !datatypeObject.Value.Equals(string.Empty))
+            {
+                hasValue = true;
+            }
+
+            var t = datatypeElement.OwnerDocument.CreateTextNode(datatypeObject.Value);
+            if (hasValue)
+            {
+                try
+                {
+                    datatypeElement.AppendChild(t);
+                }
+                catch (Exception e)
+                {
+                    throw new DataTypeException("DOMException encoding Primitive: ", e);
+                }
+            }
+
+            return hasValue;
+        }
+
+        /// <summary> Encodes a Composite in XML by looping through it's components, creating new
+        /// children for each of them (with the appropriate names) and populating them by
+        /// calling encode(Type, Element) using these children.  Returns true if at least
+        /// one component contains a value.
+        /// </summary>
+        private bool EncodeComposite(IComposite datatypeObject, XmlElement datatypeElement)
+        {
+            var components = datatypeObject.Components;
+            var hasValue = false;
+            for (var i = 0; i < components.Length; i++)
+            {
+                var name = MakeElementName(datatypeObject, i + 1);
+                var newNode = datatypeElement.OwnerDocument.CreateElement(name);
+                var componentHasValue = Encode(components[i], newNode);
+                if (componentHasValue)
+                {
+                    try
+                    {
+                        datatypeElement.AppendChild(newNode);
+                    }
+                    catch (Exception e)
+                    {
+                        throw new DataTypeException("DOMException encoding Composite: ", e);
+                    }
+
+                    hasValue = true;
+                }
+            }
+
+            return hasValue;
+        }
+
+        private void ParseReps(ISegment segmentObject, XmlElement segmentElement, string fieldName, int fieldNum)
+        {
+            var reps = segmentElement.GetElementsByTagName(fieldName);
+            for (var i = 0; i < reps.Count; i++)
+            {
+                Parse(segmentObject.GetField(fieldNum, i), (XmlElement)reps.Item(i));
+            }
+        }
+
+        /// <summary> Parses an XML element into a Varies by determining whether the element is primitive or
+        /// composite, calling setData() on the Varies with a new generic primitive or composite as appropriate,
+        /// and then calling parse again with the new Type object.
+        /// </summary>
+        private void ParseVaries(Varies datatypeObject, XmlElement datatypeElement)
+        {
+            // figure out what data type it holds

```

Comment:
```csharp
// figure out what data type it holds
// short nodeType = datatypeElement.getFirstChild().getNodeType();
datatypeObject.Data =
    !HasChildElement(datatypeElement) 
    // it's a primitive
    ? new GenericPrimitive(datatypeObject.Message)
    // it's a composite ... almost know what type, except that we don't have the version here
    : new GenericComposite(datatypeObject.Message);


Parse(datatypeObject.Data, datatypeElement);
```

Reply:
i thnk that's less readable :) - also, you'd need blank lines above the comments to satisfy the stylecop rules, and this doesnt compile with c# 7.3

Reply:
ternary statements don't work in c# 7.3? 🤯

Reply:
its not the ternary, its the generic assignment (i think) - i thnk you might need a cast on one of them